runtime/shared_common/OSCache.cpp

```diff
@@ -739,20 +741,77 @@ SH_OSCache::getAllCacheStatistics(J9JavaVM* vm, const char* ctrlDirName, UDATA g
 
 	/* Walk all the cache files in the j9shmem dir and (optionally) the persistent cacheDir */
 	do {
-		if (getCacheStatistics(vm, ctrlDirName, (const char*)nameWithVGen, groupPerm, localVerboseFlags, j2seVersion, &tempInfo, reason) != -1) {
+		J9Pool* lowerLayerList = NULL;
+		J9Pool** lowerLayerListPtr = &lowerLayerList;
+		bool isToplayer = isTopLayerCache(vm, ctrlDirName, nameWithVGen);
+		bool getLowerLayerStats = false;
+
+		if (getCacheStatistics(vm, ctrlDirName, (const char*)nameWithVGen, groupPerm, localVerboseFlags, j2seVersion, &tempInfo, reason, getLowerLayerStats, isToplayer, lowerLayerListPtr, NULL) != -1) {
 			if (includeOldGenerations || !tempInfo.isCompatible || (getCurrentCacheGen() == tempInfo.generation)) {
 				if (tempInfo.isCompatible) {
 					if (!ignoreCompatible) {
-						newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
-						memcpy(newElement, &tempInfo, sizeof(SH_OSCache_Info));
+						if (isToplayer) {
+							/* nameWithVGen is a compatible top layer cache, lower layer stats are in lowerLayerList*/
+							newElement = (SH_OSCache_Info*) pool_newElement(cacheList);
+							if (NULL == newElement) {
+								Trc_SHR_OSC_getAllCacheStatistics_pool_newElement_failed();
```

Comment:
I suggest passing in a value that can be used to identify in which location the failure occurred. Either a string or an int that is printed in the tracepoint.

Reply:
Added an int to print in the tracepoint